service/storage_service.cc

```diff
@@ -985,15 +985,14 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
             on_internal_error_noexcept(slogger, format("endpoint={} is not marked for removal but owns no tokens", endpoint));
         }
 
-        // Update pending ranges after update of normal tokens immediately to avoid
-        // a race where natural endpoint was updated to contain node A, but A was
-        // not yet removed from pending endpoints
         if (!is_normal_token_owner) {
-            auto dc_rack = get_dc_rack_for(endpoint);
-            slogger.debug("handle_state_normal: update_topology: endpoint={} dc={} rack={}", endpoint, dc_rack.dc, dc_rack.rack);
-            tmptr->update_topology(endpoint, std::move(dc_rack));
             do_notify_joined = true;
         }
+
+        // Update pending ranges after update of normal tokens immediately to avoid
+        // a race where natural endpoint was updated to contain node A, but A was
+        // not yet removed from pending endpoints
+        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
```

Comment:
How can this race happen? I don't have an understanding of why it should happen or why it shouldn't happen, but should we not serialize the callers?

Otherwise it's easy to imagine two conflicting updates getting reordered and then the first one overwriting the second.

Reply:
Wow, I merely copied this comment around and I'm not sure even if it's still valid.
It was added by @asias back in 879a4ad4d309231db11c00fc04de0da21d78de01

Reply:
This comments was moved up to the wrong place when tmptr->update_topology was added. 

There is no race now since we use the copy-update-replace method to update token_metadada.

Reply:
cool, so I'll add a patch to remove the comment then